File: src/shared/lut/pack-vault.ts
/**
 * The vault, LIVE — one module-level list of the packs this browser holds,
 * read by every picker, and the one place a pack look's lattice is resolved.
 *
 * The shape is `develop/use-preset-book.ts`'s: module state, a subscriber set,
 * loaded once per tab the first time anyone asks. A pack is not a document a
 * tool opens and closes — it is a library the whole suite reads — so it lives
 * here rather than in one screen's state.
 *
 * Resolution is cached like the built-ins' (`restore-grade.ts`): the PROMISE,
 * not the result, so a deck whose five pictures all wear the same purchased
 * look decodes it once. A 65³ lattice is 1.57 MB and decodes in ~3 ms, but
 * five of them held at once would be 5 MB of floats for nothing.
 *
 * A pack KEPT on a Winnow (`pack-remote.ts`) reaches this module in one
 * place: a `getStoredLattice` miss asks the instance, stores what comes back
 * and decodes it. That is the whole of "fetch on first use and cache" — a
 * phone that has never seen the pack grades with it the moment a picture asks,
 * and never again downloads it.
 */

import type { CubeLut } from '../lib/cube-parser';
import {
  lookIn,
  lookLabel,
  type LutPackIndex,
  type PackLook,
  type PackRef,
} from './lut-pack';
import { decodeLattice } from './pack-codec';
import {
  fetchRemoteLattice,
  fetchRemotePacks,
  packHost,
  packHosts,
  pushPack,
  type PackHost,
  type PushPackResult,
  type PushProgress,
} from './pack-remote';
import {
  deleteStoredPack,
  getStoredLattice,
  listStoredPacks,
  putStoredLattice,
  putStoredPack,
} from './pack-store';

let packs: LutPackIndex[] = [];
let loaded = false;
let loading: Promise<void> | null = null;
const listeners = new Set<() => void>();

function emit(next: LutPackIndex[]): void {
  packs = next;
  for (const fn of listeners) fn();
}

/** The packs as they stand — `[]` until the first load answers. */
export function packsSnapshot(): LutPackIndex[] {
  return packs;
}

export function subscribePacks(fn: () => void): () => void {
  listeners.add(fn);
  void loadPacks();
  return () => {
    listeners.delete(fn);
  };
}

/** Read the vault once per tab; later calls answer from memory. */
export async function loadPacks(): Promise<LutPackIndex[]> {
  if (loaded) return packs;
  loading ??= listStoredPacks().then((stored) => {
    loaded = true;
    loading = null;
    emit(stored);
  });
  await loading;
  return packs;
}

/** Tests, and the screen that empties the vault. */
export function resetPackState(): void {
  packs = [];
  loaded = false;
  loading = null;
  lattices.clear();
}

/** Add or replace a pack's index. */
export async function savePack(index: LutPackIndex): Promise<boolean> {
  const ok = await putStoredPack(index);
  emit([...packs.filter((p) => p.id !== index.id), index].sort((a, b) =>
    (a.name || a.author).localeCompare(b.name || b.author),
  ));
  return ok;
}

/** Forget a pack and its lattices. A grade that wore one of its looks then says so. */
export async function removePack(packId: string): Promise<void> {
  const pack = packs.find((p) => p.id === packId);
  await deleteStoredPack(packId);
  for (const look of pack?.looks ?? []) if (look.hash) lattices.delete(look.hash);
  emit(packs.filter((p) => p.id !== packId));
}

/** Store one look's encoded lattice — the import's write, per look. */
export async function saveLookLattice(
  packId: string,
  hash: string,
  bytes: Uint8Array,
): Promise<boolean> {
  return putStoredLattice(hash, packId, bytes);
}

/** Hide or show a node or a look in the pickers; the looks stay stored either way. */
export async function setPackHidden(packId: string, hidden: readonly string[]): Promise<void> {
  const pack = packs.find((p) => p.id === packId);
  if (!pack) return;
  await savePack({ ...pack, hidden: [...hidden] });
}

/* ------------------------------------------------------------- resolution */

const lattices = new Map<string, Promise<CubeLut | null>>();

/** The pack a reference names, if this browser holds it. */
export function packOf(ref: PackRef): LutPackIndex | null {
  return packs.find((p) => p.id === ref.pack) ?? null;
}

/** The look a reference names, if this browser holds its pack. */
export function lookOf(ref: PackRef): PackLook | null {
  const pack = packOf(ref);
  return pack ? lookIn(pack, ref.look) : null;
}

/** How a pack look is named in a stack: pack · category · camera · look. */
export function packLookName(ref: PackRef): string | null {
  const pack = packOf(ref);
  const look = pack ? lookIn(pack, ref.look) : null;
  return pack && look ? lookLabel(pack, look) : null;
}

/**
 * The lattice a reference names, decoded — or null when this device does not
 * hold it. **Null is a state, not a failure**: the layer stays in the stack
 * and says so (`restore-grade.ts`), because a purchased look missing from one
 * device must be visible rather than silently neutral.
 *
 * The HASH is what is asked for, not the look's id: it is what the bytes are
 * keyed on, and it is what catches a pack whose author published a new
 * version of the same file.
 */
export function resolvePackLattice(ref: PackRef): Promise<CubeLut | null> {
  const hash = ref.hash || lookOf(ref)?.hash || '';
  if (!hash) return Promise.resolve(null);
  const known = lattices.get(hash);
  if (known) return known;
  const pending = getStoredLattice(hash)
    .then((bytes) => bytes ?? fetchAndKeep(ref, hash))
    .then((bytes) => (bytes ? decodeLattice(bytes, packLookName(ref) ?? undefined) : null));
  // A miss is not remembered: the pack may be imported a moment later, and a
  // remembered null would keep the look missing for the rest of the session.
  pending
    .then((cube) => {
      if (!cube && lattices.get(hash) === pending) lattices.delete(hash);
    })
    .catch(() => lattices.delete(hash));
  lattices.set(hash, pending);
  return pending;
}

/**
 * The vault does not hold it: ask the instance the pack is kept on, and keep
 * what comes back. A pack that is local-only, an instance that is not
 * connected, or a network that is down all answer null — which the caller
 * reads as "not in this browser's vault", the honest state.
 */
async function fetchAndKeep(ref: PackRef, hash: string): Promise<Uint8Array | null> {
  const pack = packOf(ref);
  const host = pack?.sourceId ? packHost(pack.sourceId) : null;
  // The instance knows the bytes by their OWN hash (`PackLook.blob`), not by
  // the `.cube`'s. It is written into the index by the push, so a look whose
  // blob is blank is one this pack has never had pushed — nothing to ask for,
  // and asking under the wrong id would 404 on every picture.
  const blob = pack?.looks.find((l) => l.hash === hash)?.blob;
  if (!pack || !host || !blob) return null;
  try {
    const bytes = await fetchRemoteLattice(host, blob);
    if (!bytes) return null;
    // Cached on the way through: the phone downloads a look once, then
    // grades with it offline (`docs/lut-packs.md` §4.2 — the cache is a
    // cache, the instance is the truth).
    await saveLookLattice(pack.id, hash, bytes);
    return bytes;
  } catch {
    // Offline, signed out, or refused: the look is simply not here yet.
    return null;
  }
}

/* --------------------------------------------------------------- keeping */

/** The instances that can keep a pack — connected, with both buckets. */
export function packKeepers(): PackHost[] {
  return packHosts();
}

/**
 * Keep a pack on an instance: push its lattices, then its index, then record
 * WHERE it is kept so another device — and this one, after a cache eviction —
 * knows who to ask. The author's gesture, never automatic.
 */
export async function keepPackOn(
  packId: string,
  sourceId: string,
  onProgress?: (progress: PushProgress) => void,
): Promise<PushPackResult> {
  const pack = packs.find((p) => p.id === packId);
  const host = packHost(sourceId);
  if (!pack) throw new Error('That pack is not in this browser.');
  if (!host) throw new Error('That instance cannot keep a pack — reconnect it and try again.');
  const result = await pushPack({ ...host }, pack, (hash) => getStoredLattice(hash), onProgress);
  // The index the push WROTE, not the one it was given: it carries each look's
  // blob, which is what this device reads back to fetch a look it has evicted
  // and what the next push reads to know there is nothing to send.
  await savePack({ ...result.index, sourceId });
  return result;
}

/** The packs an instance holds that this browser does not. */
export async function remotePacksNotHere(host: PackHost): Promise<LutPackIndex[]> {
  const here = new Set(packs.map((p) => p.id));
  const there = await fetchRemotePacks(host);
  return there.filter((p) => !here.has(p.id));
}

/**
 * Take a remote pack into this browser: its INDEX only. The lattices follow
 * one at a time, as pictures ask for them — 40 MB on a phone is not something
 * to download because a list was opened.
 */
export async function adoptRemotePack(index: LutPackIndex, sourceId: string): Promise<void> {
  await savePack({ ...index, sourceId });
}

/** Why a look cannot grade here, in the words the panel shows. */
export function missingLookReason(ref: PackRef): string {
  const pack = packOf(ref);
  if (!pack) return 'This look comes from a pack this browser does not hold.';
  if (pack.sourceId && !packHost(pack.sourceId)) {
    return `Kept on ${pack.sourceId} — connect it to grade with this look.`;
  }
  return 'This look is not in this browser’s vault yet.';
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Lut/PackVault.swift
// The VAULT — the packs this device holds, read by every picker, and the one
// place a pack look's lattice is resolved. Port of `src/shared/lut/pack-vault.ts`
// over the pure half of `pack-store.ts`; the storage itself (the web's
// IndexedDB database `atelier-lut-packs`) is the app's, handed in as a
// `PackStore`, and the instances as `PackHost`s (`PackRemote.swift`).
//
// The web keeps this as MODULE state with a subscriber set, because a pack is
// a library the whole suite reads rather than a document one screen opens.
// Here it is an ACTOR the app holds one of: the same verbs, the same
// interleaving at each `await` that the web's promises have.
//
// Rules kept (`docs/lut-packs.md`, `media-pipeline.md`):
// - Resolution is cached by HASH and caches the pending work, not the result,
//   so five pictures wearing one purchased look decode it once. A MISS is
//   never remembered: the pack may be imported a moment later.
// - A lattice this device lacks is asked of the instance its pack is kept on
//   — under the look's `blob`, never its `hash` — and stored on the way
//   through: fetched on first use, then graded with offline (§4.2 — the
//   instance is the truth, this vault its cache).
// - A pack kept on an instance is forgotten THERE or not at all: an
//   unreachable instance refuses the gesture and changes nothing, and the
//   remote half runs BEFORE the local write, so a failure leaves both whole.
//   Only the lattices NO surviving look names are freed (`freedHashes`).
// - Every gesture that writes to an instance is the author's (`keepPackOn`,
//   `adoptRemotePack`); nothing here writes there on its own.

import Foundation

/// Why a vault gesture was refused — the web's thrown message, verbatim.
public struct PackVaultError: Error, Equatable, Sendable, CustomStringConvertible {
    public let message: String
    public var description: String { message }

    public init(_ message: String) { self.message = message }
}

/// The vault's storage on this device — what the web's `pack-store.ts` does
/// over IndexedDB. Two stores: one INDEX per pack (by pack id) and one ENCODED
/// lattice per look (`PackCodec.swift` bytes, never `.cube` text), keyed by the
/// SHA-256 of the SOURCE `.cube`, so two packs shipping one file store it once.
///
/// Every call DEGRADES rather than throws, as the web's do: a device may deny
/// or evict storage, and a missing lattice has a defined meaning here (the
/// look says it is not in this vault). A write answers false when it failed;
/// an empty hash is never stored nor found.
public protocol PackStore: Sendable {
    /// Every pack held, read through `readStoredPacks`; `[]` when unusable.
    func listStoredPacks() async -> [LutPackIndex]
    func putStoredPack(_ index: LutPackIndex) async -> Bool
    /// Forget an INDEX. Its lattices are `deleteStoredLattices`' — which ones
    /// are free is a question about every index at once, asked by the vault.
    func deleteStoredPack(_ packId: String) async
    func deleteStoredLattices(_ hashes: [String]) async
    func getStoredLattice(_ hash: String) async -> [UInt8]?
    func putStoredLattice(_ hash: String, packId: String, bytes: [UInt8]) async -> Bool
    /// Which hashes this device holds — what a re-import skips.
    func storedLatticeHashes() async -> Set<String>
    /// What each stored lattice weighs, by hash — the MEASURED half of
    /// `PackWeight.swift`; empty when it cannot say, never zeros.
    func storedLatticeSizes() async -> LatticeSizes
}

/// Packs in the order every list shows them: by name (the author's, for a
/// pack with none), as the web's `localeCompare` sorts, stable on a tie.
public func sortPacks(_ packs: [LutPackIndex]) -> [LutPackIndex] {
    packs.enumerated().sorted { a, b in
        let ka = a.element.name.isEmpty ? a.element.author : a.element.name
        let kb = b.element.name.isEmpty ? b.element.author : b.element.name
        let order = packLocaleCompare(ka, kb)
        return order != 0 ? order < 0 : a.offset < b.offset
    }.map(\.element)
}

/// The stored rows read back — `pack-store.ts`'s `listStoredPacks` less the
/// IndexedDB: each migrated, a row that is not an index left behind, sorted.
public func readStoredPacks(_ rows: [JSONValue]) -> [LutPackIndex] {
    sortPacks(rows.compactMap(migratePackIndex))
}

private func vaultText(_ s: String?) -> String? {
    guard let s, !s.isEmpty else { return nil }
    return s
}

/// The vault, live. One per app; every picker reads it.
public actor PackVault {
    private let store: any PackStore
    /// The instances that can keep a pack, as they stand now — the web's
    /// `packHosts()`, read on every ask because a connection comes and goes.
    private let hosts: @Sendable () -> [PackHost]

    private var packs: [LutPackIndex] = []
    private var loaded = false
    private var loading: Task<Void, Never>?
    private var listeners: [(id: Int, fn: @Sendable () -> Void)] = []
    private var nextListener = 0
    private var lattices: [String: Task<CubeLut?, Never>] = [:]

    public init(store: any PackStore, hosts: @escaping @Sendable () -> [PackHost] = { [] }) {
        self.store = store
        self.hosts = hosts
    }

    private func emit(_ next: [LutPackIndex]) {
        packs = next
        for listener in listeners { listener.fn() }
    }

    /// The packs as they stand — `[]` until the first load answers.
    public func packsSnapshot() -> [LutPackIndex] { packs }

    /// Be told when the packs change; starts the first load. The id unsubscribes.
    public func subscribePacks(_ fn: @escaping @Sendable () -> Void) -> Int {
        nextListener += 1
        listeners.append((nextListener, fn))
        Task { _ = await self.loadPacks() }
        return nextListener
    }

    public func unsubscribePacks(_ id: Int) {
        listeners.removeAll { $0.id == id }
    }

    /// Read the vault once per session; later calls answer from memory.
    public func loadPacks() async -> [LutPackIndex] {
        if loaded { return packs }
        let pending: Task<Void, Never>
        if let loading {
            pending = loading
        } else {
            pending = Task {
                let stored = await store.listStoredPacks()
                self.loaded = true
                self.loading = nil
                self.emit(sortPacks(stored))
            }
            loading = pending
        }
        await pending.value
        return packs
    }

    /// Tests, and the screen that empties the vault.
    public func resetPackState() {
        packs = []
        loaded = false
        loading = nil
        lattices.removeAll()
    }

    /// Add or replace a pack's index.
    @discardableResult
    public func savePack(_ index: LutPackIndex) async -> Bool {
        let ok = await store.putStoredPack(index)
        emit(sortPacks(packs.filter { $0.id != index.id } + [index]))
        return ok
    }

    /// Store one look's encoded lattice — the import's write, per look.
    public func saveLookLattice(_ packId: String, _ hash: String, _ bytes: [UInt8]) async -> Bool {
        await store.putStoredLattice(hash, packId: packId, bytes: bytes)
    }

    /// Which lattices this device holds, by source hash.
    public func storedLatticeHashes() async -> Set<String> {
        await store.storedLatticeHashes()
    }

    /// What each stored lattice weighs — for `packWeight` / `vaultWeight`.
    public func storedLatticeSizes() async -> LatticeSizes {
        await store.storedLatticeSizes()
    }

    /// Hide or show a node or a look in the pickers; the looks stay stored.
    public func setPackHidden(_ packId: String, _ hidden: [String]) async {
        guard var pack = packs.first(where: { $0.id == packId }) else { return }
        pack.hidden = hidden
        await savePack(pack)
    }

    // MARK: forgetting

    /// Forget a whole pack: its index and its lattices, here and on the
    /// instance it is kept on. A grade that wore one of its looks then says so.
    public func removePack(_ packId: String) async throws -> ForgetResult {
        try await forget(packId, nil)
    }

    /// Forget ONE look — reclaiming its bytes, where hiding only puts it away.
    public func forgetLook(_ packId: String, _ lookId: String) async throws -> ForgetResult {
        try await forget(packId, [lookId])
    }

    /// `lookIds == nil` means the whole pack.
    private func forget(_ packId: String, _ lookIds: [String]?) async throws -> ForgetResult {
        _ = await loadPacks()
        guard let pack = packs.first(where: { $0.id == packId }) else {
            throw PackVaultError("That pack is not in this browser.")
        }
        let whole = lookIds == nil
        let doomed = lookIds ?? pack.looks.map(\.id)
        let keptOn = pack.sourceId
        let host = vaultText(keptOn).flatMap { hostFor($0) }
        if let keptOn = vaultText(keptOn), host == nil {
            throw PackVaultError(
                "This pack is kept on \(keptOn). Connect it, so \(whole ? "the pack" : "the look") "
                    + "is forgotten there too and its bytes come back."
            )
        }

        let sizes = await store.storedLatticeSizes()
        let here = freedHashes(packs, packId, doomed)
        let there = host.map { freedBlobs(packs, packId, doomed, $0.sourceId) } ?? Freed()
        let next = whole ? nil : withoutLooks(pack, doomed)
        // Weighed BEFORE anything is written: the write takes the very looks
        // both figures are read from out of the indexes.
        let result = ForgetResult(
            here: hashBytes(here.free, sizes),
            instance: host != nil ? blobBytes(there.free, packs) : 0,
            keptOn: keptOn,
            shared: here.shared.count
        )

        if let host {
            if let next {
                try await deleteRemoteLooks(host, next, there.free)
            } else {
                try await deleteRemotePack(host, pack, Set(there.shared))
            }
        }

        if let next {
            await savePack(next)
        } else {
            await store.deleteStoredPack(packId)
            emit(packs.filter { $0.id != packId })
        }
        await store.deleteStoredLattices(here.free)
        // The decode cache is keyed on the hash: a look whose bytes are gone
        // must lose its entry or the session keeps grading with it.
        for hash in here.free { lattices[hash] = nil }
        return result
    }

    // MARK: resolution

    private func hostFor(_ sourceId: String) -> PackHost? {
        packHost(sourceId, in: hosts())
    }

    /// The pack a reference names, if this device holds it.
    public func packOf(_ ref: PackRef) -> LutPackIndex? {
        packs.first { $0.id == ref.pack }
    }

    /// The look a reference names, if this device holds its pack.
    public func lookOf(_ ref: PackRef) -> PackLook? {
        packOf(ref).flatMap { lookIn($0, ref.look) }
    }

    /// How a pack look is named in a stack: pack · category · camera · look.
    public func packLookName(_ ref: PackRef) -> String? {
        guard let pack = packOf(ref), let look = lookIn(pack, ref.look) else { return nil }
        return lookLabel(pack, look)
    }

    /// The lattice a reference names, decoded — or nil when this device does
    /// not hold it. **Nil is a state, not a failure**: the layer stays in the
    /// stack and says so (`restoreLayers`' `.missing`). The HASH is what is
    /// asked for, not the look's id: it is what the bytes are keyed on, and
    /// what catches a pack whose author published a new version of a file.
    public func resolvePackLattice(_ ref: PackRef) async -> CubeLut? {
        let hash = vaultText(ref.hash) ?? vaultText(lookOf(ref)?.hash) ?? ""
        if hash.isEmpty { return nil }
        if let known = lattices[hash] { return await known.value }
        let pending = Task { () -> CubeLut? in
            var bytes = await self.store.getStoredLattice(hash)
            if bytes == nil { bytes = await self.fetchAndKeep(ref, hash) }
            guard let bytes else { return nil }
            return decodeLattice(bytes, title: self.packLookName(ref))
        }
        lattices[hash] = pending
        let cube = await pending.value
        // A miss is not remembered: the pack may be imported a moment later.
        if cube == nil, lattices[hash] == pending { lattices[hash] = nil }
        return cube
    }

    /// The vault does not hold it: ask the instance the pack is kept on, under
    /// the look's `blob`, and keep what comes back. Local-only, not connected,
    /// never pushed, offline — all answer nil, "not in this vault".
    private func fetchAndKeep(_ ref: PackRef, _ hash: String) async -> [UInt8]? {
        guard let pack = packOf(ref), let sourceId = vaultText(pack.sourceId), let host = hostFor(sourceId),
              let blob = vaultText(pack.looks.first(where: { $0.hash == hash })?.blob) else { return nil }
        do {
            guard let bytes = try await fetchRemoteLattice(host, blob) else { return nil }
            _ = await saveLookLattice(pack.id, hash, bytes)
            return bytes
        } catch {
            return nil
        }
    }

    // MARK: keeping

    /// The instances that can keep a pack — connected, with both buckets.
    public func packKeepers() -> [PackHost] { hosts() }

    /// Keep a pack on an instance: push its lattices, then its index, then
    /// record WHERE it is kept — the index the push WROTE, carrying each look's
    /// blob, which is what this device reads back to fetch an evicted look.
    public func keepPackOn(_ packId: String, _ sourceId: String,
                           onProgress: ((PushProgress) -> Void)? = nil) async throws -> PushPackResult {
        let pack = packs.first { $0.id == packId }
        let host = hostFor(sourceId)
        guard let pack else { throw PackVaultError("That pack is not in this browser.") }
        guard let host else {
            throw PackVaultError("That instance cannot keep a pack — reconnect it and try again.")
        }
        let store = self.store
        let result = try await pushPack(host, pack, latticeFor: { await store.getStoredLattice($0) },
                                        onProgress: onProgress)
        var kept = result.index
        kept.sourceId = sourceId
        await savePack(kept)
        return result
    }

    /// The packs an instance holds that this device does not.
    public func remotePacksNotHere(_ host: PackHost) async throws -> [LutPackIndex] {
        let here = Set(packs.map(\.id))
        return try await fetchRemotePacks(host).filter { !here.contains($0.id) }
    }

    /// Take a remote pack in: its INDEX only. The lattices follow one at a
    /// time as pictures ask — 40 MB is not downloaded because a list was opened.
    public func adoptRemotePack(_ index: LutPackIndex, _ sourceId: String) async {
        var adopted = index
        adopted.sourceId = sourceId
        await savePack(adopted)
    }

    /// Why a look cannot grade here, in the words the panel shows.
    public func missingLookReason(_ ref: PackRef) -> String {
        guard let pack = packOf(ref) else { return "This look comes from a pack this browser does not hold." }
        // The pack is here and this look is NOT: forgotten, or republished
        // without it — "not here yet" would be a lie about a thing dropped.
        if lookIn(pack, ref.look) == nil {
            let name = !pack.name.isEmpty ? pack.name : (!pack.author.isEmpty ? pack.author : "this pack")
            return "That look is no longer in \(name) — forgotten, or gone from the pack."
        }
        if let sourceId = vaultText(pack.sourceId), hostFor(sourceId) == nil {
            return "Kept on \(sourceId) — connect it to grade with this look."
        }
        return "This look is not in this browser’s vault yet."
    }
}
